validator/db/kv/attestation_history_new.go

```diff
@@ -0,0 +1,144 @@
+package kv
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	bolt "go.etcd.io/bbolt"
+	"go.opencensus.io/trace"
+)
+
+// the size of each data entry in bytes source epoch(8 bytes) signing root (32 bytes)
+const uint64Size = 8
+const latestEpochWrittenSize = uint64Size
+const targetSize = uint64Size
+const sourceSize = uint64Size
+const signingRootSize = 32
+const historySize = targetSize + sourceSize + signingRootSize
+const minimalSize = latestEpochWrittenSize
+
+// AttestationHistoryNew stores the historical attestation data needed
+// for protection of validators.
+type AttestationHistoryNew struct {
+	TargetToSource     map[uint64]*HistoryData
+	LatestEpochWritten uint64
+}
+
+// HistoryData stores the needed data to confirm if an attestation is slashable
+// or repeated.
+type HistoryData struct {
+	Source      uint64
+	SigningRoot []byte
+}
+
+func newAttestationHistoryArray(target uint64) []byte {
+	enc := make([]byte, latestEpochWrittenSize+(target%params.BeaconConfig().WeakSubjectivityPeriod)*historySize+historySize)
+	return enc
+}
+
+func sizeChecks(enc []byte) error {
```

Comment:
Does `enc` need to be passed in here? How about just its length int?

Reply:
+1. We don't need `enc` inside the func, but `len(enc)`, so better just pass `encSize int` as a parameter.

Alternatively (please use your own judgement whether this is needed), we can also introduce a separate type, and consolidate all these funcs within scope of that type as receiver? That's:
```golang
type historyData []byte // or attestationHistoryArray, or sth else -- whatever is more descriptive

func (hd historyData) assertSize() error { // pointer receiver will also work
  if (len(hd)-minimalSize)%historySize != 0 {
    return fmt.Errorf("encapsulated data size: %d is not a multiple of entry size: %d", len(hd), historySize)
  }
}

// Then w/i client code:
data := historyData(raw bytes) // type conversion
if err := data.assertSize(); err != nil {
  // do sth
}
//  Or, assuming `newAttestationHistoryArray` returns `historyData`:
data := newAttestationHistoryArray()
```

I believe, that, overall, we should strive to limit number of funcs that are not methods i.e. funcs that do not have a corresponding receiver, funcs that are just dumped into global/package scope. When we bind funcs with receivers, we scope them, and normally end up with a more clean API as well.